### PR TITLE
publisher/ deviceServer

### PR DESCRIPTION
**Pull request by AntonPuko**

### CODE REVIEW

File: src/clients/SparkCore.js [hunk 1]
```diff
@@ -68,7 +68,6 @@ const MAX_BINARY_SIZE = 108000; // According to the forums this is the max size.
  */
 class SparkCore extends EventEmitter {
   _cipherStream: ?Duplex = null;
-  _connectionKey: ?string = null;
```
Comment (AntonPuko) [inline src/clients/SparkCore.js:71]: connectionKey is redundant..its used only in some console.logs inside sparkCore and in DeviceServer for checking if the device is in devicesByIdMap...

File: src/lib/EventPublisher.js [hunk 2]
```diff
@@ -61,37 +72,51 @@ class EventPublisher extends EventEmitter {
...
-      eventHandler,
+      listener,
     } = nullthrows(this._subscriptionsByID.get(subscriptionID));
 
-    this.removeListener(eventName, eventHandler);
+    this.removeListener(eventName, listener);
     this._subscriptionsByID.delete(subscriptionID);
   };
+
+  unsubscribeBySubscriberID = (subscriberID: string): void => {
```
Comment (AntonPuko) [inline src/lib/EventPublisher.js:112]: add subscriberID to be able to clean all listeners for the current device if it is disconnected.

File: src/server/DeviceServer_v2.js [hunk 3]
```diff
@@ -263,6 +267,54 @@ class DeviceServer {
     await this._eventPublisher.publish(eventData);
   };
 
+  _onDeviceSubscribe = async (
+    message: Message,
+    device: SparkCore,
+  ): Promise<void> => {
+    const deviceID = device.getHexCoreID();
+    // uri -> /e/?u    --> firehose for all my devices
```
Comment (AntonPuko) [inline src/server/DeviceServer_v2.js:302]: this info is very messy..I understand nothing from it, do I need to remove?

File: src/server/DeviceServer_v2.js [hunk 4]
```diff
@@ -263,6 +267,54 @@ class DeviceServer {
...
+    // uri -> /e/event_name?u (deviceid)    --> deviceid?
+    const messageName = message.getUriPath().substr(3);
+
+    if (!messageName) {
+      device.sendReply('SubscribeFail', message.getId());
+      return;
+    }
+
+    const query = message.getUriQuery();
+    const fromMyDevicesOnly = query && !!query.match('u');
```
Comment (AntonPuko) [inline src/server/DeviceServer_v2.js]: not sure about the name of this var. what is this for:  it checks if the device wants to listen event from devices with the same ownerID as it has. `Particle.subscribe( lall, lulu, MY_DEVICES) ` - in firmware

File: src/lib/EventPublisher.js [hunk 5]
```diff
@@ -61,37 +72,51 @@ class EventPublisher extends EventEmitter {
...
+    this._subscriptionsByID
+      .forEach((subscription) => {
```
Comment (AntonPuko) [inline src/lib/EventPublisher.js:114]: native js Map() doesn't have filter()..didn't now that before.

File: src/clients/SparkCore.js (hunk 1 above)
Comment (jlkalberer) [reply src/clients/SparkCore.js:71]: Yeah, it is convenient for the logs though... Maybe just have a global counter and use that instead.

File: src/clients/SparkCore.js [hunk 6]
```diff
@@ -188,21 +187,16 @@ class SparkCore extends EventEmitter {
     logger.log(
       'On Device Ready:\r\n',
       {
-        cache_key: this._connectionKey,
```
Comment (jlkalberer) [inline src/clients/SparkCore.js:191]: This still needs to be there. It helps me when debugging.

File: src/clients/SparkCore.js [hunk 7]
```diff
@@ -921,8 +915,7 @@ class SparkCore extends EventEmitter {
     logger.error(
       'This client has an exclusive lock',
       {
-        cache_key: this._connectionKey,
```
Comment (jlkalberer) [inline src/clients/SparkCore.js:924]: ditto

File: src/lib/EventPublisher.js (hunk 5 above)
Comment (jlkalberer) [reply src/lib/EventPublisher.js:114]: This seems odd... Why don't you:
```
const subscription = this._subscriptionsByID.get(subscriptionID)
if (subscription) {
  this.unsubscribe(subscription.id);
}
```

File: src/server/DeviceServer_v2.js (hunk 3 above)
Comment (jlkalberer) [reply src/server/DeviceServer_v2.js:302]: Nope, please keep it.  It's the COAP protocol

File: src/server/DeviceServer_v2.js (hunk 4 above)
Comment (jlkalberer) [reply src/server/DeviceServer_v2.js]: `isFromMyDevices`

**Review by jlkalberer — CHANGES_REQUESTED** [review]

**Review by jlkalberer — APPROVED** [review]

**Pull request merged**